renderer/tr_light.cpp

```diff
@@ -1007,6 +1008,7 @@ it and any necessary overlays
 ===================
 */
 idRenderModel *R_EntityDefDynamicModel( idRenderEntityLocal *def ) {
+	std::lock_guard<std::mutex> lock (def->mutex);
```

Comment:
C++11 threading spreads over the code...

Meahwhile, there are two in-house libraries for multithreading. The one from original Doom 3 is pretty limited (4 numbered critical sections), but the one which came from Doom 3 BFG supports creating arbitrary number of mutexes.

Why not use `idSysMutex` and `idScopedCriticalSection` then?

Reply:
Wouldn't it be better to use the std::mutex and get rid of idSysMutex, instead? Not only is it less code to maintain for us, but idSysMutex uses a CRITICAL_SECTION on Windows, and since VS 2017 (or even 2015?) the std::mutex implementation uses SRWLock, which should be faster.

Reply:
I have a feeling that it's better to change underlying implementation of `idSysMutex`, if there is something faster than CriticalSection. I tried to find something about the performance difference, and the first argument was: SRWLock is 8 bytes instead of 40 bytes of critical section. Ha-ha, std::mutex is 48 bytes =)

By the way, SRW sounds like reader-writer lock. Shouldn't it be something that works good when there are many read-only users? If yes, then it needs different interface, I guess.

Reply:
Oh, std::mutex is 80 bytes in 64-bit mode =)

Reply:
But why is it better? It would mean additional coding effort, and again, code that we'd have to maintain and test across our platforms, instead of a solution that we get out of the box. And sure, std::mutex has the usual bloat of STL, but if we ever use so many mutexes that that's going to be an issue, we'll have other problems :)

SRWLocks, according to their documentation, have two modes of operation. One is indeed many readers/one writer, the other is a single exclusive access. So I guess MS crammed two different concepts into the same structure :D

Reply:
No coding effort is necessary. Just replace `std::mutex` with `idSysMutex`, and everything will continue working. The performance difference is hypothetical both ways =)

Mixing different frameworks does not sound well. BFG code already uses idSys primitives. I have already used a few to make idFileSystem thread-safe.

One nasty question I had recently is compatibility of C++11 with pthreads. Some time ago we had problems with Linux build hanging or crashing on exit. At that moment I tried to search for an answer to this simple question, but surprisingly did not find any.

And on the point of bragging, me and Duzenko had bad experience with `std::condition_variable`. In my opinion it is too low-level to be used by itself and too easy to misuse. Win32 events make more sense, which we have in form of idSysSignal.

Reply:
Fine, I've changed the two mutexes to idSysMutex.

But honestly, I think your disdain for the STL and simultaneous reverance of id code sometimes borders on religious ;)
The only reason we even have idSysMutex is because duzenko copied thousands of lines of code from BFG, which is code that is completely untested outside of its BFG habitat, and I've already had to modify it to sensibly support more than 2 worker threads. There were C++ task libraries (building on top of C++14 threading primitives) available that could have done the same job for us without now having to be responsible for that BFG threading code, and I think they would have been the smarter choice. If we ever find the need to establish a new platform (or even reactivate macOS, although with OpenGL being deprecated, I guess we'll have avoided that one), this notion of relying on id platform code is going to bite us in the ass...

Reply:
Thank you!

I'm not exactly "fond" of the ID code, to be honest. The `idList` container is pretty weird, bears too little resemblance to STL to be intuitive, and its original code even lacked some things like exponentially growing push_back. But we have it, and it works.

As for STL... yes, I dislike it a lot. Call it religion if you like. The previous TDM lead seemed to have boost religion BTW =) We can start a holywar on the topic and have a lot of fun =)

One major point against STL is debug performance. It just sucks, for no apparent reason. And just like build time, debug performance does not drop dramatically every time you use STL. It goes down gradually and without notice. No particular commit can be blamed for it, the effect of one small change can't be measured. But in the end STL is everywhere, and debug performance sucks, and everyone is sure this is normal. On my daily job I don't care much, because I can record the calls to my code, and replay it without any external code. But for TDM debug performance is very important.

I have recently wrote a [small article about debug performance](https://dirtyhandscoding.github.io/posts/fast-debug-in-visual-c.html) (it even mentions TDM). And I guess you won't be surprised if I say that I am writing a replacement for some bits of STL (don't worry, it won't get into TDM). And its vector works much faster than the one from STL in debug. E.g. if I only enable \Ob1, it is 5 times faster on the benchmark from the article. It is the inherent cost of STL, it will never go away.

Anecdotical cases include the story about you using `std::regex` in shader manager. Its debug performance is so bad that it really hurts!
The sad fact is that when files are packaged into pk4 files, the TDM packager uses regexes to decide what to include and what not, and which pk4 to put it into. Packaging TDM in debug takes hours! I am happy that I don't need to debug it often.